sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/aio/_link_async.py

```diff
@@ -244,6 +244,9 @@ async def _incoming_detach(self, frame) -> None:
             self._error = error_cls(condition=frame[2][0], description=frame[2][1], info=frame[2][2])
             await self._set_state(LinkState.ERROR)
         else:
+            if self.state != LinkState.DETACH_SENT:
+                raise AMQPLinkError(condition=ErrorCondition.InternalError,
```

Comment:
Would this be better as an UnknownError, since we don't actually get the InternalError condition from the service until we try attaching again after closing the connection?

Reply:
I agree - InternalError doesn't seem appropriate here.
Additionally - I don't think we should be raising in this method at all - given that this is a handler for any incoming frame - we have no way to predict where this error would ultimately surface and therefore where the error handling needs to live. It should be using the same assignment to `self._error` as above on line 244 (this mechanism has its own problems - but it's the status quo for now.....).
Also - raising like this will skip the `set_state` line below as well.

Additionally - are we 100% sure that a service initiated DETACH without error condition should always be an error? Are there any scenarios where a service DETACH would be normal, or at least expected behaviour (including in EventHubs)?

Reply:
@annatisch I checked with some of the other languages and how they handle this situation in their libraries

java - treats this as a retriable event in their stack
.net - handles the detach and end up closing their link and connection. When the next operation happens, they realize that the connection is closed and re-open everything up ( in the process retrying )
go - also treats such a detach as retriable. They do a recovery at the link level itself. 

This sort of detach happens during a swap, service upgrade etc. and the libraries treat it "why does the service want to close, try again".


For us, by setting `self._error` in link, we cause `client_ready_async` to raise an error from `connection.listen`